### PR TITLE
win_dns_record

### LINKED ISSUE #51923: FeatReq: Add support for DNS record management on Windows DNS Server
##### SUMMARY
Add a `win_dns_record` module for managing DNS records in Windows DNS server (ie usually extra records in an AD environment).

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME

win_dns_record (proposed)

##### ADDITIONAL INFORMATION

Active Directory has unique dependencies on DNS -- so much so that it requires its own, dedicated DNS system to work effectively. For some organizations, it may even be their primary DNS platform. Thus support for managing records within that system is sought.

### PR DESCRIPTION
**Pull request by johnboy2**
##### SUMMARY

Fixes ansible/ansible#51923

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

win_dns_record

##### ADDITIONAL INFORMATION
n/a

### CODE REVIEW

**Comment by angystardust** [comment]
Hi @johnboy2 and thanks for this very useful module!
I've done some easy tests with ansible 2.7.5 and a Win2k12R2:
- adding an A or a CNAME record works fine but there's no idempotency: every time I run the play, the module emits a `changed` status instead of an `ok` :disappointed: 
- removing an A or a CNAME record works fine and the idempotency is satisfied :+1:

**Comment by johnboy2** [comment]
@angystardust - I believe I have that idempotency issue fixed.

**Comment by it-praktyk** [comment]
The pull request don't contain integration tests :-(

**Comment by angystardust** [comment]
@johnboy2 great! I'll take another "test tour" tomorrow ;)

**Comment by angystardust** [comment]
So, I've done some tests again. Now idempotency is enforced for an A record but still gives a "changed" status for exiting CNAME records

**Comment by johnboy2** [comment]
@Daniel-Sanchez-Fabregas - FYI I had to revert 9cf5f2d, as it breaks check-mode by omitting most of the (simulated) changes. Maybe that approach can be brought back in if we can find a better way to simulate the records.

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 1]
```diff
@@ -0,0 +1,148 @@
+#!powershell
+
+# Copyright: (c) 2019, Hitachi ID Systems, Inc.
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#Requires -Module Ansible.ModuleUtils.Legacy
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]: This module is still using the legacy Powershell module interface.
We do have a new Powershell module interface which has many improvements.
Would you care to rewrite this module using the new interface?
Please look at some other Windows modules for examples.

File: lib/ansible/modules/windows/win_dns_record.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    - The name of the zone to manage (eg C(example.com)). The zone must already
+      exist.
+    required: yes
+    type: str
+  computer_name:
+    description:
+      - Specifies a DNS server. You can specify an IP address or any value that
+        resolves to an IP address, such as a fully qualified domain name
+        (FQDN), host name, or NETBIOS name.
+    required: false
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.py]: If the parameter is not required, there is no need to add this.
```suggestion
```

File: lib/ansible/modules/windows/win_dns_record.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: win_dns_record
+version_added: "2.8"
+short_description: Manage Windows Server DNS records
+description:
+- Manages DNS records within an existing Windows Server DNS zone
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.py]: All descriptions, except the short_description, must end with a trailing dot.
```suggestion
- Manages DNS records within an existing Windows Server DNS zone.
```

File: lib/ansible/modules/windows/win_dns_record.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    type: str
+  state:
+    description:
+    - Whether the record should exist or not.
+    choices: [ absent, present ]
+    default: present
+    type: str
+  ttl:
+    description:
+    - The "time to live" of the record, in seconds. Ignored when
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.py]: Use multiple lines. The first line explains the parameter briefly.
Subsequent lines clarify when to use it, how to use it, what values are expected etc.
```suggestion
    - The "time to live" of the record, in seconds.
    - Ignored when
```

File: lib/ansible/modules/windows/win_dns_record.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    required: yes
+    type: str
+  value:
+    description:
+    - The value(s) to specify. Required when C(state=present).
+    aliases: [ values ]
+    type: str
+  zone:
+    description:
+    - The name of the zone to manage (eg C(example.com)). The zone must already
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.py]:
```suggestion
    - The name of the zone to manage (eg C(example.com)).
    - The zone must already
```

File: lib/ansible/modules/windows/win_dns_record.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      exist.
+    required: yes
+    type: str
+  computer_name:
+    description:
+      - Specifies a DNS server. You can specify an IP address or any value that
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.py]:
```suggestion
      - Specifies a DNS server.
      - You can specify an IP address or any value that
```

**Review by dagwieers — CHANGES_REQUESTED** [review]
LGTM, although we may want to improve it further.

File: test/integration/targets/win_dns_record/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,19 @@
+- name: Clean slate
+  import_tasks: clean.yml
+  vars:
+    fail_on_missing: false
+
+- block:
```
Comment (Daniel-Sanchez-Fabregas) [inline test/integration/targets/win_dns_record/tasks/main.yml]: Trying to replicate the tests locally breaks for me as DNS feature is missing. This solve it to me:
~~~ yaml
   - name: check if feature is availble
     win_shell: if (Get-Command -Name Add-WindowsFeature -ErrorAction SilentlyContinue) { $true } else { $false }
     changed_when: False
     register: module_available

   - name: install DNS features
     when: module_available.stdout | trim | bool
     block:

       - name: ensure DNS services are installed
         win_feature:
           name: DNS
           state: present
         register: dns_install

   - name: reboot server if needed
     win_reboot:
     when: dns_install.reboot_required
~~~

Comment (Daniel-Sanchez-Fabregas) [reply test/integration/targets/win_dns_record/tasks/main.yml]: See https://github.com/johnboy2/ansible/pull/2

File: test/integration/targets/win_dns_record/tasks/tests-PTR.yml [hunk 8]
```diff
@@ -0,0 +1,186 @@
+- name: 'TYPE=PTR - creation (check mode)'
+  win_dns_record: {zone: '{{ win_dns_record_revzone }}', name: 7, value: ansible-mirror.example.com, type: PTR}
+  register: cmd_result
+  check_mode: yes
+
+- name: 'TYPE=PTR - creation get results (check mode)'
+  win_command: powershell.exe "If (Get-DnsServerResourceRecord -Zone '{{ win_dns_record_revzone }}' -Name '7' -RRType PTR -Node -ErrorAction:Ignore) { 'exists' } else { 'absent' }"
```
Comment (Daniel-Sanchez-Fabregas) [inline test/integration/targets/win_dns_record/tasks/tests-PTR.yml]: In order to make the tests work for meI have to substitute in this file all instances of `-Zone` by `-ZoneName` 
See https://github.com/johnboy2/ansible/pull/2

File: lib/ansible/modules/windows/win_dns_record.ps1 (hunk 1 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_dns_record.ps1]: This was written by following the [developer guide](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general_windows.html#windows-new-module-development), which still recommends the legacy module. My bias is to always studiously avoid stepping outside the documented box -- meaning I would prefer waiting for the developer guide to show some new path before jumping in.

Comment (varunchopra) [reply lib/ansible/modules/windows/win_dns_record.ps1]: You're correct to go by the docs however that page has been updated for Ansible 2.8 and you can view it [here](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general_windows.html#windows-new-module-development).

The new module interface was introduced in the latest developmental release and would be great if you could use this instead. You wouldn't have to change a lot of code to use it.

Comment (Daniel-Sanchez-Fabregas) [reply lib/ansible/modules/windows/win_dns_record.ps1]: Tried to convert it in https://github.com/johnboy2/ansible/pull/2/commits/6a7d92a0791e5d58d11cd410d3a4c45d549820b7 . Tests are OK after conversion, but don know how to:
* ~~use `prepared` in `diff`~~
* use `required_if` in this context

Comment (Daniel-Sanchez-Fabregas) [reply lib/ansible/modules/windows/win_dns_record.ps1]: I have this solved in https://github.com/johnboy2/ansible/pull/2 
Seems to me that `required_if` don't apply in this case.
Tests are OK

**Comment by johnboy2** [comment]
@dagwieers - Your review points were applied, but remain outstanding in the UI. Can you dismiss and/or re-review?

**Comment by angystardust** [comment]
@dagwieers @johnboy2 no chance to see this one in 2.8 GA?

**Comment by Daniel-Sanchez-Fabregas** [comment]
Maybe 951164ec4915e318f30e63b02c7cff83a5345fd7 was needed https://app.shippable.com/github/ansible/ansible/runs/112013/17/tests
@johnboy2 could you try to pass the tests adding it?

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+$spec = @{
+    options = @{
+        name = @{ type = "str"; required = $true }
+        state = @{ type = "str"; choices = "absent", "present"; default = "present" }
+        ttl = @{ type = "int"; default = "3600" }
+        type = @{ type = "str"; choices = "A","AAAA","CNAME","PTR"; required = $true }
+        value = @{ type = "list"; elements = "str"; default = @() ; aliases=@( 'values' )}
+        zone = @{ type = "str"; required = $true }
+        computer_name = @{ type = "str"; required = $false }
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]: If it is not required, it is not needed.
```suggestion
        computer_name = @{ type = "str" }
```

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+$state = $module.Params.state
+$ttl = $module.Params.ttl
+$type = $module.Params.type
+$values = $module.Params.value
+$zone = $module.Params.zone
+$dns_computer_name = $module.Params.computer_name
+
+
+$extra_args = @{}
+if ($dns_computer_name -ne $null) {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]: Thanks to PowerShell's design...
```suggestion
if ($null -ne $dns_computer_name) {
```

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+
+$changes = @{
+    before = "";
+    after = ""
+}
+
+
+$records = Get-DnsServerResourceRecord -ZoneName $zone -Name $name -RRType $type -Node -ErrorAction:Ignore @extra_args | Sort-Object
+if ($records -ne $null) {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]:
```suggestion
if ($null -ne $records) {
```

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+
+$extra_args = @{}
+if ($dns_computer_name -ne $null) {
+    $extra_args.ComputerName = $dns_computer_name
+}
+
+if ($state -eq 'present') {
+    if ($values.Count -eq 0) {
+        $module.FailJson("values must be non-empty when state='present'")
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]: Capitalize error messages. (You have some capitalized in this script, so keep it consistent)

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    }
+} else {
+    if ($values.Count -ne 0) {
+        $module.FailJson("values must be undefined or empty when state='absent'")
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]: Capitalize.

File: lib/ansible/modules/windows/win_dns_record.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = r'''
+---
+module: win_dns_record
+version_added: "2.8"
+short_description: Manage Windows Server DNS records
+description:
+- Manages DNS records within an existing Windows Server DNS zone.
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.py]:
```suggestion
- Manage DNS records within an existing Windows Server DNS zone.
```

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+
+# TODO: add warning for forest minTTL override -- see https://docs.microsoft.com/en-us/windows/desktop/ad/configuration-of-ttl-limits
+if ($ttl -lt 1 -or $ttl -gt 31557600) {
+    $module.FailJson("ttl must be between 1 and 31557600")
+}
+$ttl = New-TimeSpan -Seconds $ttl
+
+
+if (($type -eq 'CNAME' -or $type -eq 'PTR') -and $values -ne $null -and $values.Count -gt 0 -and $zone[-1] -ne '.') {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]:
```suggestion
if (($type -eq 'CNAME' -or $type -eq 'PTR') -and $null -ne $values -and $values.Count -gt 0 -and $zone[-1] -ne '.') {
```

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+            $module.Result.changed = $true
+        }
+    }
+
+    # Whatever is left in $required_values needs to be added
+    $values = $required_values.Keys
+}
+
+
+if ($values -ne $null -and $values.Count -gt 0) {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_dns_record.ps1]:
```suggestion
if ($null -ne $values -and $values.Count -gt 0) {
```

**Review by dagwieers — CHANGES_REQUESTED** [review]
LGTM, a few changes.

I would also like to see full support for all record types, but if we can start with only a subset, that's fine for me too.

**Comment by dagwieers** [comment]
@angystardust I hope so. It would help if more people review and test the code, and add a shipit.

**Comment by dagwieers** [comment]
The tests seem to fail on Windows 2008 and 2008R2, so probably should be excluded from running on those.

**Comment by angystardust** [comment]
@dagwieers thanks for your review. I'm already using the module in the current status on a 2.7 setup and..."it works for me" :copyright: 
Happy to test further if a newer revision is pushed ;)

File: lib/ansible/modules/windows/win_dns_record.ps1 (hunk 12 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_dns_record.ps1]: I wasn't sure how to deal with that one given that "values" is an input variable, but "Values" is not.

File: lib/ansible/modules/windows/win_dns_record.ps1 (hunk 13 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_dns_record.ps1]: Ditto my last.

File: lib/ansible/modules/windows/win_dns_record.ps1 (hunk 12 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_dns_record.ps1]: Actually I'm just going to reword it so the first word isn't the attribute name.

Comment (johnboy2) [reply lib/ansible/modules/windows/win_dns_record.ps1]: I've reworked the affected sentences instead.

**Comment by johnboy2** [comment]
> Maybe [951164e](https://github.com/ansible/ansible/commit/951164ec4915e318f30e63b02c7cff83a5345fd7) was needed https://app.shippable.com/github/ansible/ansible/runs/112013/17/tests
> @johnboy2 could you try to pass the tests adding it?

My concerns with that idea are that it leaves the feature uninstalled if it won't work, and that I shouldn't have to work around functionality already provided by (and presumably tested for) the `win_feature` module.

A few other things:
 * @dagwieers' suggestions applied. For the capitalization ones, the affected sentences have been reworked.
 * RE failing tests, looks like Win2008/etc don't support the essential cmdlets used by this module. So I've also added it to the skip list for those platforms (also @dagwieers' suggestion) plus added that fact to the module documentation.
 * RE supporting other record types, I agree completely but didn't have time to even think about other types, much less test them, consider corner cases, etc. For me at least, priority <span>#</span>1 was getting something workable out the door.

File: lib/ansible/modules/windows/win_dns_record.ps1 [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    }
+} else {
+    if ($values.Count -ne 0) {
+        $module.FailJson("Parameter 'values' must be undefined or empty when state='absent'")
+    }
+}
+
+
+# TODO: add warning for forest minTTL override -- see https://docs.microsoft.com/en-us/windows/desktop/ad/configuration-of-ttl-limits
+if ($ttl -lt 1 -or $ttl -gt 31557600) {
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_dns_record.ps1:49]: This should be documented in the option description.

File: test/integration/targets/win_dns_record/tasks/main.yml [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
```
Comment (jborean93) [inline test/integration/targets/win_dns_record/tasks/main.yml]: While you've added the skip in the aliases, you should still check and skip if necessary when running on older versions of Windows. Usually it's a simple

```
- win_shell: if (Get-Command -Name ... -ErrorAction SilentlyContinue) { $true } else { $false }
  register: run_tests

- name: run tests on supported hosts
  when: run_tests.stdout | trim | bool
  block:
```

**Review by jborean93 — CHANGES_REQUESTED** [review]
Looks great, don't know much about DNS but you've covered pretty much all of the basics.

File: lib/ansible/modules/windows/win_dns_record.ps1 (hunk 17 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_dns_record.ps1:49]: Agreed -- I'll add that range limit.

File: test/integration/targets/win_dns_record/tasks/main.yml (hunk 18 above)
Comment (johnboy2) [reply test/integration/targets/win_dns_record/tasks/main.yml]: Alright, I know when I'm beat (@Daniel-Sanchez-Fabregas mentioned something similar).

But keep in mind that the relevant cmdlets aren't available _even on supported platforms_ unless/until the DNS feature is installed.

So I've added the cmdlet tests, but also put OS-version tests in front of them to avoid installing the DNS feature unnecessarily. That has the effect of making the cmdlet test redundant, however.

File: lib/ansible/modules/windows/win_dns_record.ps1 (hunk 17 above)
Comment (johnboy2) [reply lib/ansible/modules/windows/win_dns_record.ps1:49]: Done.

File: test/integration/targets/win_dns_record/tasks/main.yml (hunk 18 above)
Comment (jborean93) [reply test/integration/targets/win_dns_record/tasks/main.yml]: Ah sorry that makes sense, I would just do the OS check then, sorry for the confusion.

Comment (johnboy2) [reply test/integration/targets/win_dns_record/tasks/main.yml]: Done.

File: test/integration/targets/win_dns_record/tasks/tests.yml [hunk 19]
```diff
@@ -1,3 +1,13 @@
+- name: ensure DNS services are installed
+  win_feature:
+    name: DNS
+    state: present
+  register: dns_install
+
+- name: reboot server if needed
+  win_reboot:
+  when: dns_install.reboot_required
```
Comment (Daniel-Sanchez-Fabregas) [inline test/integration/targets/win_dns_record/tasks/tests.yml:9]: @johnboy2. Maybe better Asking for changes here? like:

https://github.com/ansible/ansible/blob/0e980596376be7b7ef843d20b52228687e94717d/test/integration/targets/win_dns_record/tasks/main.yml#L16-L18

Comment (Daniel-Sanchez-Fabregas) [reply test/integration/targets/win_dns_record/tasks/tests.yml:9]: Ooops, you just changed it. :innocent: Though this was other file.

Comment (johnboy2) [reply test/integration/targets/win_dns_record/tasks/tests.yml:9]: :-)

**Comment by johnboy2** [comment]
@jborean93 @dagwieers - can you re-review?

**Review by jborean93 — APPROVED** [review]
Looks great, thanks for working on this one.
